File: solutions/typescript/all-your-base/1/all-your-base.ts
export function convert(
  digits: number[],
  inputBase: number,
  outputBase: number 
): number[]{
  switch (true) {
    case inputBase <= 1: throw Error("Wrong input base")
    case outputBase <= 1: throw Error("Wrong output base")
    case outputBase % 1 !== 0: throw Error("Wrong output base")
    case inputBase % 1 !== 0: throw Error("Wrong input base")
    case digits.length === 0: throw Error("Input has wrong format")
    case digits.some(digit => digit < 0): throw Error("Input has wrong format")
    case digits.some(digit => digit >= inputBase): throw Error("Input has wrong format")
    case digits.length > 1 && digits[0] === 0: throw Error("Input has wrong format")
  }

  return toBase(fromBase(digits, inputBase), outputBase)

}

function fromBase(digits: number[], base: number): number {
  return digits.reverse().reduce((acc, digit, index) => {
    return acc + digit * base ** index
  }, 0)

}

function toBase(number: number, base: number): number[] {
  if (number === 0)  {
    return [0]
  }
  let output = []
  while (number > 0) {
    output.push(number % base)
    number = Math.floor(number / base)
  }

  return output.reverse()
}

File: solutions/c/all-your-base/1/all_your_base.c
#include "all_your_base.h"

size_t rebase(int8_t digits[], int16_t input_base, int16_t output_base, size_t input_length) {
    if (input_base < 2 || output_base < 2) {
        return 0; // Invalid base
    }
    if (digits == NULL || input_length == 0) {
        return 0; // Invalid input
    }
    for (size_t i = 0; i < input_length; i++) {
        if (digits[i] < 0 || digits[i] >= input_base) {
            return 0; // Invalid digit
        }
    }
    // Convert input digits to decimal
    int32_t value = 0;
    for (size_t i = 0; i < input_length; i++) {
        value = value * input_base + digits[i];
    }
    if (value == 0) {
        digits[0] = 0;
        return 1; // Output is zero
    }
    // Convert decimal value to output base
    int8_t output_digits[DIGITS_ARRAY_SIZE];
    size_t output_length = 0;
    while (value > 0) {
        if (output_length >= DIGITS_ARRAY_SIZE) {
            return 0; // Output exceeds maximum length
        }
        output_digits[output_length++] = value % output_base;
        value /= output_base;
    }
    // Reverse the output digits
    for (size_t i = 0; i < output_length; i++) {
        digits[i] = output_digits[output_length - 1 - i];
    }
    return output_length;

}
